protocols/autonatv2/src/client/handler.rs

```diff
@@ -0,0 +1,48 @@
+pub(crate) mod dial_back;
+pub(crate) mod dial_request;
+
+use std::{
+    fmt::{Display, Formatter},
+    sync::Arc,
+    time::Duration,
+};
+
+pub(crate) use dial_request::TestEnd;
+
+use self::dial_request::InternalError;
+
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);
+const MAX_CONCURRENT_REQUESTS: usize = 10;
+
+#[derive(Clone, Debug)]
```

Comment:
That is necessary because of this line: https://github.com/umgefahren/rust-libp2p/blob/be35e80000db92188a8991bd4caa0fc6aea02c47/protocols/autonatv2/src/client/handler/dial_request.rs#L275 

However I will change it to not using the trait, since it ends up in the public API

Let me illustrate the problem in a bigger picture. I'll try to show the rough flow of an error from the start to where it is handled:
```mermaid
flowchart TD
handle_substream[[handle_substream]]
subgraph start_substream_handle
handle_substream --> res
res --> status_update{{StatusUpdate}}
end

subgraph handler_poll
res --> test_completed{{ToBehaviour::TestCompleted}}
test_completed --> notify_behaviour{{ConnectionHandlerEvent::NotifyBehaviour}}
end

subgraph behaviour [Behaviour::on_connection_handler_event]
notify_behaviour --> get_error_cause{Get the cause of the error}
get_error_cause -- Other Error --> tracing[[tracing::debug]]
get_error_cause -- FailureDuringDialBack / UnableToConnectOnSelectedAddress --> expired{{ToSwarm::ExternalAddrExpired}}

status_update --> to_swarm{{ToSwarm::GenerateEvent}}
end
```

Essentially I needed a way to share a single error across the two uses shown above. The Arc and the clone was the easiest solution. The Arc is sadly necessary, because `std::io::Error` doesn't implement `Clone`.

I know it's complicated and I'm not entirely happy with the solution either. But it makes it possible to isolate the two runs well and the performance penalty is negligible.

If you are feeling strongly about this, I will change it.

Reply:
Thank you for the thorough explanation! Looking at the code, it appears to me that this design was influenced from a time when we would send multiple `StatusUpdate`s. Now that we are only reported the total bytes sent, couldn't we just get rid of `StatusUpdate` and return the information in a single flow as part of the return value of `start_substream_handle`?

I that appears to be a simpler design and we don't need to clone the error then.

Reply:
Ok, we could keep the StatusUpdate though. Just inject the Result later when we matched on the Error.

Reply:
Changed it